library/alloc/src/ffi/c_str.rs

```diff
@@ -690,6 +693,53 @@ impl CString {
             }),
         }
     }
+
+    /// Attempts to convert a <code>[Vec]<[u8]></code> to a [`CString`].
+    ///
+    /// The input [`Vec`] must contain at least one nul byte.
+    ///
+    /// If the nul byte is not at the end of the input `Vec`, then the `Vec`
+    /// will be truncated so that there is only one nul byte, and that byte
+    /// is at the end.
+    ///
+    /// # Errors
+    ///
+    /// If no nul byte is present, an error will be returned.
+    ///
+    /// # Examples
+    /// ```
+    /// #![feature(cstr_from_bytes_until_nul)]
+    ///
+    /// use std::ffi::CString;
+    ///
+    /// let mut buffer = vec![0u8; 16];
+    /// unsafe {
+    ///     // Here we might call an unsafe C function that writes a string
+    ///     // into the buffer.
+    ///     let buf_ptr = buffer.as_mut_ptr();
+    ///     buf_ptr.write_bytes(b'A', 8);
+    /// }
+    /// // Attempt to extract a C nul-terminated string from the buffer.
+    /// let c_str = CString::from_vec_until_nul(buffer).unwrap();
+    /// assert_eq!(c_str.into_string().unwrap(), "AAAAAAAA");
+    /// ```
+    ///
+    #[unstable(feature = "cstr_from_bytes_until_nul", issue = "95027")]
+    pub fn from_vec_until_nul(mut v: Vec<u8>) -> Result<Self, FromVecWithNulError> {
+        let nul_pos = memchr::memchr(0, &v);
+        match nul_pos {
+            Some(nul_pos) => {
+                v.truncate(nul_pos + 1);
+                // SAFETY: We know there is a nul byte at nul_pos, so this slice
+                // (ending at the nul byte) is a well-formed C string.
+                Ok(unsafe { Self::_from_vec_with_nul_unchecked(v) })
+            }
+            None => Err(FromVecWithNulError {
```

Comment:
I would rather a separate error type as a wrapper around FromBytesUntilNulError, since FromBytesWithNullErrorKind::InteriorNul shouldn't be valid

Reply:
Sorry, I didn't understand what you mean by "as a wrapper around..."
Are you suggesting something like this?

```rust
struct FromVecUntilNulError {
    inner: FromVecWithNulError
}

impl FromVecUntilNulError {
    // ... proxy `as_bytes`, `into_bytes` to the inner type
}
```

As opposed to the "plain version" with no wrapper:

```rust
struct FromVecUntilNulError {
    bytes: Vec<u8>,
}

impl FromVecUntilNulError {
    // ... copy/paste `as_bytes`, `into_bytes` code from FromVecWithNulError
}
```

Can you help me understand your suggestion?

Reply:
I was thinking something like this,

```rs
struct FromVecUntilNulError {
    inner: FromBytesUntilNulError,
    bytes: Vec<u8>
}
```

But I since saw that FromBytesUntilNulError is a zst, so your suggestion of a 'plain version' with no wrapper seems good.

Reply:
OK, added the plain `FromVecUntilNulError`.